File: go/bingo_board_bdd_test.go
package bingo_board_test

import (
	"testing"

	bingo_board "github.com/sammancoaching/Bingo-Refactoring-Kata"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestBDDANewlyCreatedBoardIsNotInitialised(t *testing.T) {
	testCase := bingoBoardBDDTestCase{}
	testCase.givenABingoBoardOfSize(1, 1)
	testCase.thenTheBoardIsNotInitialised(t)
}

func TestBDDWhenAllFieldsAreSetTheBoardIsInitialised(t *testing.T) {
	anyValue := "42"

	testCase := bingoBoardBDDTestCase{}
	testCase.givenABingoBoardOfSize(1, 1)
	require.NoError(t, testCase.whenCellIsDefined(0, 0, anyValue))
	testCase.thenTheBoardIsInitialised(t)
}

func TestBDDWhenAllFieldsOnRectangularBoardAreSetItIsInitialised(t *testing.T) {
	require := require.New(t)
	one := "one, two, three"
	two := "Bingo cells can contain any text"

	testCase := bingoBoardBDDTestCase{}
	testCase.givenABingoBoardOfSize(1, 2)
	require.NoError(testCase.whenCellIsDefined(0, 0, one))
	require.NoError(testCase.whenCellIsDefined(0, 1, two))
	testCase.thenTheBoardIsInitialised(t)
}

func TestBDDADefinedCellCantBeRedefinedEvenIfItsTheSameValue(t *testing.T) {
	anyValue := "42"

	testCase := bingoBoardBDDTestCase{}
	testCase.givenABingoBoardOfSize(1, 1)
	require.NoError(t, testCase.whenCellIsDefined(0, 0, anyValue))
	assert.Error(t, testCase.whenCellIsDefined(0, 0, anyValue))
}

func TestBDDDuplicateCellsAreNotAllowed(t *testing.T) {
	anyValue := "42"

	testCase := bingoBoardBDDTestCase{}
	testCase.givenABingoBoardOfSize(2, 2)
	require.NoError(t, testCase.whenCellIsDefined(0, 1, anyValue))
	assert.Error(t, testCase.whenCellIsDefined(1, 0, anyValue))
}

func TestBDDANonInitialisedBoardCannotBeMarked(t *testing.T) {
	require := require.New(t)
	anyValue := "42"

	testCase := bingoBoardBDDTestCase{}
	testCase.givenABingoBoardOfSize(1, 1)
	require.NoError(testCase.whenCellIsDefined(0, 0, anyValue))
	require.NoError(testCase.whenCellIsMarked(0, 0))
	testCase.thenTheCellIsMarked(t, 0, 0)
}

type bingoBoardBDDTestCase struct {
	board bingo_board.BingoBoard
}

func (c *bingoBoardBDDTestCase) givenABingoBoardOfSize(width, height int) {
	c.board = bingo_board.New(width, height)
}

func (c *bingoBoardBDDTestCase) whenCellIsDefined(x, y int, value string) error {
	return c.board.DefineCell(x, y, value)
}

func (c *bingoBoardBDDTestCase) whenCellIsMarked(x, y int) error {
	return c.board.MarkCell(x, y)
}

func (c *bingoBoardBDDTestCase) thenTheBoardIsNotInitialised(t *testing.T) {
	assert.False(t, c.boardInitialisedState())
}

func (c *bingoBoardBDDTestCase) thenTheBoardIsInitialised(t *testing.T) {
	assert.True(t, c.boardInitialisedState())
}

func (c *bingoBoardBDDTestCase) boardInitialisedState() bool {
	return c.board.IsInitialised()
}

func (c *bingoBoardBDDTestCase) thenTheCellIsMarked(t *testing.T, x, y int) {
	assert.True(t, c.board.IsMarked(x, y))
}

File: ruby/bingo_board_bdd_test.rb
# frozen_string_literal: true

require 'minitest/autorun'
require_relative 'bingo_board'

class BingoBoardBDDTest < Minitest::Test
  def test_a_newly_created_board_is_not_initialized
    given_a_bingo_board_of_size(1, 1)
    then_the_board_is_not_initialized
  end

  def test_when_all_fields_are_set_the_board_is_initialized
    any_value = '42'

    given_a_bingo_board_of_size(1, 1)
    when_cell_is_defined(0, 0, any_value)
    then_the_board_is_initialized
  end

  def test_when_all_fields_on_rectangular_board_are_set_it_is_initialized
    one = 'one, two, three'
    two = 'Bingo cells can contain any text'

    given_a_bingo_board_of_size(1, 2)
    when_cell_is_defined(0, 0, one)
    when_cell_is_defined(0, 1, two)
    then_the_board_is_initialized
  end

  def test_a_defined_cell_cant_be_redefined_even_if_its_the_same_value
    any_value = '42'

    given_a_bingo_board_of_size(1, 1)
    when_cell_is_defined(0, 0, any_value)
    
    assert_raises(BingoBoard::IllegalStateException) do
      when_cell_is_defined(0, 0, any_value)
    end
  end

  def test_duplicate_cells_are_not_allowed
    any_value = '42'

    given_a_bingo_board_of_size(2, 2)
    when_cell_is_defined(0, 1, any_value)
    
    assert_raises(BingoBoard::IllegalStateException) do
      when_cell_is_defined(1, 0, any_value)
    end
  end

  def test_a_cell_can_be_marked_on_initialized_board
    any_value = '42'

    given_a_bingo_board_of_size(1, 1)
    when_cell_is_defined(0, 0, any_value)
    when_cell_is_marked(0, 0)
    then_the_cell_is_marked(0, 0)
  end

  def test_a_non_initialized_board_cannot_be_marked
    given_a_bingo_board_of_size(1, 1)
    
    assert_raises(BingoBoard::IllegalStateException) do
      when_cell_is_marked(0, 0)
    end
  end

  private

  def given_a_bingo_board_of_size(width, height)
    @board = BingoBoard.new(width, height)
  end

  def when_cell_is_defined(x, y, value)
    @board.define_cell(x, y, value)
  end

  def when_cell_is_marked(x, y)
    @board.mark_cell(x, y)
  end

  def then_the_board_is_not_initialized
    refute board_initialized_state
  end

  def then_the_board_is_initialized
    assert board_initialized_state
  end

  def board_initialized_state
    @board.initialized?
  end

  def then_the_cell_is_marked(x, y)
    assert @board.marked?(x, y)
  end
end